pkg/kubelet/kuberuntime/kuberuntime_container_linux.go

```diff
@@ -51,23 +52,14 @@ func (m *kubeGenericRuntimeManager) generateLinuxContainerConfig(container *v1.C
 	}
 
 	// set linux container resources
-	var cpuShares int64
 	cpuRequest := container.Resources.Requests.Cpu()
 	cpuLimit := container.Resources.Limits.Cpu()
 	memoryLimit := container.Resources.Limits.Memory().Value()
 	oomScoreAdj := int64(qos.GetContainerOOMScoreAdjust(pod, container,
 		int64(m.machineInfo.MemoryCapacity)))
 	// If request is not specified, but limit is, we want request to default to limit.
```

Comment:
According to this comment, the current behaviour is correct: if the request is 0 and the limit is set, cpuShares will default to the limit. A workload cannot request 0 CPU; that is equivalent to setting no request. If you want request to be 2 CPU shares, set that explicitly.

/hold

Reply:
It's very strange. 

for example:

set request to 0 explicitly

```
resources:
  limits:
    cpu: 2
    memory: 1G
  requests:
    cpu: 0
    memory: 0
```

qos: **burstable**  cpushares: 2048

not set requests

```
resources:
  limits:
    cpu: 2
    memory: 1G
```

qos: **guaranteed** cpushares: 2048
apiserver will set request default to limit if there is no request set explicitly and save it to etcd. In etcd it will have a requests value that equals to limits. 


Just as the comment said, If request is not specified, but limit is, we want request to default to limit. In my case even i set request to 0 explicitly, the cpushares also be set by limit.

If limit is 2c, the cpushares of container whose request set to 1c is less than that request set to 0 explicitly.

if the current behaviour is correct, how can i set container level cpushares to 2, by set request to 1m or 2m?

Reply:
Thanks for your reply. 

It's very strange. 

for example:

set request to 0 explicitly

```
resources:
  limits:
    cpu: 2
    memory: 1G
  requests:
    cpu: 0
    memory: 0
```

qos: **burstable**  cpu.shares: 2048

not set requests

```
resources:
  limits:
    cpu: 2
    memory: 1G
```

qos: **guaranteed** cpu.shares: 2048

apiserver will set request default to limit if there is no request set explicitly and save it to etcd. In etcd it will have a requests value that equals to limits. 


Just as the comment said, If request is not specified, but limit is, we want request to default to limit. In my case even i set request to 0 **_explicitly_**, the cpu.shares of container level also be set by limit. In the meantime, pod level cgroup and qos level cgroup  will not set it as limit value. 


If limit is 2c, the cpu.shares of container whose request set to 1c is less than that request set to 0 explicitly.

if the current behaviour is correct, how can i set container level cpu.shares to 2, by set request to 1m or 2m?

Reply:
> According to this comment, the current behaviour is correct: if the request is 0 and the limit is set, cpuShares will default to the limit. A workload cannot request 0 CPU; that is equivalent to setting no request. If you want request to be 2 CPU shares, set that explicitly.
> 
> /hold

Thanks for your reply. 

It's very strange. 

for example:

set request to 0 explicitly

```
resources:
  limits:
    cpu: 2
    memory: 1G
  requests:
    cpu: 0
    memory: 0
```

qos: **burstable**  cpu.shares: 2048

not set requests

```
resources:
  limits:
    cpu: 2
    memory: 1G
```

qos: **guaranteed** cpu.shares: 2048

apiserver will set request default to limit if there is no request set explicitly and save it to etcd. In etcd it will have a requests value that equals to limits. 


Just as the comment said, If request is not specified, but limit is, we want request to default to limit. In my case even i set request to 0 **_explicitly_**, the cpu.shares of container level also be set by limit. In the meantime, pod level cgroup and qos level cgroup  will not set it as limit value. 


If limit is 2c, the cpu.shares of container whose request set to 1c is less than that request set to 0 explicitly.

if the current behaviour is correct, how can i set container level cpu.shares to 2, by set request to 1m or 2m?